### ISSUE TITLE
Edges unmount and remount when linked nodes are (de)selected

### ISSUE THREAD

**Issue by Sec-ant** [wbkd/react-flow#1966]
### Describe the Bug

When selecting or deselecting a node, its connected edges will unmount and remount, which will cause the internal states in custom edges being reset.

### Your Example Website or App

https://codesandbox.io/s/eager-lewin-zxygu1

### Steps to Reproduce the Bug or Issue

1. Open the demo
2. Select or deselect `Node 5`, `Output 8` or `Output 9` and watch the internal counter value and console log.

### Expected behavior

Selecting or deselecting a node shouldn't unmount or remount its connected edges. The internal states in the custom edges should persist.

### Screenshots or Videos

https://user-images.githubusercontent.com/10386119/158066715-21a7d76b-3330-4d92-9c05-58cfda4dd479.mp4

### Platform

- OS: [Windows]
- Browser: [Edge]
- Version: [10.0.0-next.51]

**Comment by Sec-ant** [comment]
After some digging around I found the reason. The `EdgeWrapper`s are grouped into different `svg` elements depending on their different `z-index` levels. So when a node is selected, its connected edges will all have higher levels and thus are regrouped into a new svg element, whose `key` is the new level, and trigger a remount.

The `edgeTree` store data is also structured to group different edges into different levels. I might not be aware of the design considerations behind this. I managed to flatten the `EdgeWrapper`s and wrap each of them into an individual svg element to fix this issue.

https://github.com/Sec-ant/react-flow/commit/ee1cc74e2a0c6c70ca17330ab3475d0d53bea471#diff-687a1bfc2516c89b169da9501420c0cde3e47018204ea282d65597d42d7155e9

The remounting of edges will cause some visual glitches as well even if you don't save states in the edge component (note the animation and edge label in the demo below):

https://user-images.githubusercontent.com/10386119/158503073-4b430837-228b-43c7-bfac-bf0fc15d212f.mp4

Will you guys reconsider the data structure of this? @moklick Thanks!
